File: fsharp/focl/p058.fs
module Focl.P058

open Xunit
open FsUnit.Xunit

(*
https://4clojure.oxal.org/#/problem/58
Problem 58, Function Composition
Difficulty: medium
Write a function which allows you to create function compositions. 
The parameter list should take a variable number of functions, 
and create a function applies them from right-to-left.
    (= [3 2 1] ((__ rest reverse) [1 2 3 4]))
    (= 5 ((__ (partial + 3) second) [1 2 3 4]))
    (= true ((__ zero? #(mod % 8) +) 3 5 7 9))
    (= "HELLO" ((__ #(.toUpperCase %) #(apply str %) take) 5 "hello world"))
*)

let l1, r1 = [ 1; 2; 3; 4 ], [ 3; 2; 1 ]

let f1 fa fb = fun x -> x |> fb |> fa

[<Fact>]
let t_058_a () =
    let f = f1 List.tail List.rev
    f l1 |> should equal [ 3; 2; 1 ]

let f058b a b = fun x -> a (b x)

[<Fact>]
let t058b () =
    let f = f058b List.tail List.rev
    f l1 |> should equal [ 3; 2; 1 ]

File: clojure/src/ace/focl/p058.clj
(ns ace.focl.p058
  (:require
   [clojure.test :as t]))

;; Write a function which allows you to create function compositions.
;; The parameter list should take a variable number of functions,
;; and create a function applies them from right-to-left.
;; Special restriction: comp

(defn run [f]
  (and (= [3 2 1] ((f rest reverse) [1 2 3 4]))
       (= 5 ((f (partial + 3) second) [1 2 3 4]))
       (= true ((f zero? #(mod % 8) +) 3 5 7 9))
       (= "HELLO" ((f #(.toUpperCase %) #(apply str %) take) 5 "hello world"))))

(defn a [& fs]
  (fn [& vs]
    (reduce (fn [result f] (f result))
            (apply (last fs) vs)
            (reverse (butlast fs)))))

(t/deftest all
  (t/are [f] (run f)
    a))
